lib/airbrake/rack/instrumentable.rb

```diff
@@ -15,12 +17,99 @@ module Rack
+        @__airbrake_capture_timing_module__ ||= ::Module.new
+      end
+
+      # Using api private self methods so they don't get defined in the target
+      # class or module, but can still be called by the above method.
+
+      # @api private
+      def self.should_prepend?(klass, method_name)
+        # Don't chain already-prepended or operator methods.
+        klass.module_exec do
+          self_class_idx = ancestors.index(self)
+          method_owner_idx = ancestors.index(instance_method(method_name).owner)
+          method_owner_idx < self_class_idx || !(/\A\W/ =~ method_name).nil?
+        end
+      end
 
-        define_method(method_name) do |*args|
-          Airbrake::Rack.capture_timing(label) do
-            __send__("#{method_name}_without_airbrake", *args)
+      # @api private
+      def self.prepend_airbrake_capture_timing(klass, method_name, label)
+        args = method_signature
+        klass.module_exec do
+          # Check the visibility.
+          is_protected = protected_method_defined?(method_name)
+          is_private = private_method_defined?(method_name)
+          mod = __airbrake_capture_timing_module__
+          mod.module_exec do
+            # Generate the wrapper method.
+            module_eval <<-RUBY, __FILE__, __LINE__ + 1
+              def #{method_name}(#{args})
+                Airbrake::Rack.capture_timing(#{label.to_s.inspect}) do
+                  super
+                end
+              end
+            RUBY
+            # Restore the visibility.
+            protected method_name if is_protected
+            private method_name if is_private
           end
+          prepend mod
+        end
+      end
+
+      # @api private
+      def self.chain_airbrake_capture_timing(klass, method_name, label)
+        args = method_signature
+        klass.module_exec do
+          # Check the visibility.
+          is_public = private_method_defined?(method_name)
+          is_protected = protected_method_defined?(method_name)
+          is_private = private_method_defined?(method_name)
+
+          # Generate the wrapper method.
+          aliased = method_name.to_s.sub(/([?!=])$/, '')
+          punctuation = Regexp.last_match(1)
+          alias_method "#{aliased}_without_airbrake#{punctuation}", method_name
+          remove_method method_name
+          module_eval <<-RUBY, __FILE__, __LINE__ + 1
+            def #{method_name}(#{args})
+              Airbrake::Rack.capture_timing(#{label.to_s.inspect}) do
+                __send__("#{aliased}_without_airbrake#{punctuation}", #{args})
+              end
+            end
+          RUBY
+
+          # Restore the visibility.
+          public method_name if is_public
+          protected method_name if is_protected
+          private method_name if is_private
+        end
+      end
+
+      # @api private
+      # Using a method instead of a constant so it isn't defined in the target.
```

Comment:
If we use a constant, it won't be defined in the target. That said, it indeed will be available from the target's context, true.

```rb
[1] pry(main)> module M
[1] pry(main)*   FOO = 1
[1] pry(main)* end
=> 1
[2] pry(main)> class A
[2] pry(main)*   extend M
[2] pry(main)* end
=> A
[3] pry(main)> class A
[3] pry(main)*   class << self
[3] pry(main)*     p FOO
[3] pry(main)*   end
[3] pry(main)* end
1
=> 1
```

Therefore, I think we should say this:

```suggestion
      # Using a method instead of a constant so it isn't accessible in the target.
```

Reply:
Sounds good to me.